packages/connection-manager/src/js/Connection/XHRConnection.js

```diff
@@ -0,0 +1,198 @@
+    Object.defineProperty(this, "data", {
+      get: () => data,
+      set: __data => {
+        data = __data;
+      }
+    });
+    this.data = _data;
+
+    var header = {};
+    Object.defineProperty(this, "header", {
+      get: () => header,
+      set: __header => {
+        header = Object.assign(header, __header);
+      }
+    });
+    this.header = _header;
+
+    var xhr = null;
+    Object.defineProperty(this, "xhr", {
+      get: () => xhr,
+      set: _xhr => {
+        if (
+          xhr !== null &&
+          (this.state !== CONNECTION_STATE_INIT ||
+            this.state !== CONNECTION_STATE_CANCELED)
+        ) {
+          xhr.abort();
+          this.emit(CONNECTION_EVENT_ERROR);
+        }
+        xhr = _xhr;
+        this.state = CONNECTION_STATE_INIT;
+      }
+    });
+
+    this.handleOpen = this.handleOpen.bind(this);
+    this.handleProgress = this.handleProgress.bind(this);
+    this.handleAbort = this.handleAbort.bind(this);
+    this.handleError = this.handleError.bind(this);
+    this.handleLoad = this.handleLoad.bind(this);
+    this.handleTimeout = this.handleTimeout.bind(this);
+  }
+
+  /**
+   * more xhr related getter
+   * @todo make them more robust / do we need more?
+   */
+  get response() {
+    return this.xhr.response;
+  }
+  get readyState() {
+    return this.xhr.readyState;
+  }
+  get responseType() {
+    return this.xhr.responseType;
+  }
+  get status() {
+    return this.xhr.status;
+  }
+
+  /**
+   * create, prepare, open and send the xhr request
+   * @param {string} [token] – authentication token
+   */
+  open(token) {
+    if (this.xhr !== null) {
+      throw new Error("cannot open XHR Connection a second time!");
+    }
+    this.xhr = new XMLHttpRequest();
+
+    this.xhr.addEventListener("progress", this.handleProgress);
+    this.xhr.addEventListener("abort", this.handleAbort);
+    this.xhr.addEventListener("error", this.handleError);
+    this.xhr.addEventListener("load", this.handleLoad);
+    this.xhr.addEventListener("timeout", this.handleTimeout);
+    this.xhr.open(this.method, this.url);
+
+    this.handleOpen();
+
+    Object.keys(this.header).forEach(key => {
+      if (this.header[key] !== undefined && this.header[key] !== null) {
+        this.xhr.setRequestHeader(key, this.header[key]);
+      }
+    });
+
+    if (token !== undefined && token !== "") {
+      this.xhr.setRequestHeader("Authorization", "Bearer " + token);
+    }
+
+    this.xhr.send(this.data);
+  }
+  /**
+   * aborts native xhr
+   */
+  close() {
+    this.xhr.abort();
+  }
+  /**
+   *
+   * @todo abort here before delete?
+   */
+  reset() {
+    this.state = CONNECTION_STATE_CANCELED;
+    this.xhr = null;
+  }
+
+  /**
+   * handle open of native xhr
+   */
+  handleOpen() {
+    this.state = CONNECTION_STATE_STARTED;
+    this.emit(CONNECTION_EVENT_OPEN, this);
+  }
+  /**
+   * handle open  of native xhr
+   */
+  handleProgress() {
+    this.emit(CONNECTION_EVENT_PROGRESS, this);
```

Comment:
I tried to use it and got a question: why do we want to pass `this` to an event listener? Usually `ProgressEvent` goes there. Had to modify it like this to make it work:
```
handleProgress(event) {
  this.emit(CONNECTION_EVENT_PROGRESS, event);
}
```